R/modify-recursive.R

```diff
@@ -0,0 +1,54 @@
+#' Recursively modify a list
+#'
+#' `rmodify()` allows you to recursively modify a vector, supplying functions
+#' to either modify each leaf or each node (or both).
+#'
+#' @param x A list.
+#' @param f_leaf A function applied to each leaf.
+#' @param f_pre A function applied to each node or leaf, before leaves are
+#'   transformed.
```

Comment:
This actually made me thinking, is it necessary to apply `f_pre` and `f_post` to leafs? I think most commonly the `f_pre` and `f_post` will be used on nodes and in general the same operation is unlikely to be applied on both leafs and nodes. As a result a user would need to add an extra check for leafs inside `f_pre` and `f_post`. 

Also applying all 3 transformations on leafs is redundant and slightly confusing.

Reply:
I’m glad you bought that up because that’s what I was thinking too 😀

Reply:
Hmmmm, what happens if `f_pre` changes `x` from a node to a leaf?

Reply:
As a `map`-like function, I think it should preserve the types/shape of the data structure; `f_pre`/`f_post` should have `Node -> Node` type and should not delete a node or transmute a node into a leaf. Otherwise, the behavior is a bit surprising, depending on how the tree is traversed.

Reply:
I think being able to change a type of a node is a very important feature. For example one might want to change the end list nodes (prior to leafs) into a proper atomic vector. 

I find it hard to think of good use-cases for f_pre, but the one you mentioned seem an important one. It's quite possible that the purpose of an operation is to transform a sub-tree into a leaf. The traversal stops right there. This can be important for deeply nested trees where performance starts playing a role.  On the other hand `p_leaf` can capture this use case as well.  

So there are 4 options:

  1) leave as it is and add a note that `f_pre` should return a node, otherwise the behavior is undefined.
  2) add an extra check for leaf after f_pre
  3) make f_pre/f_post asymetric and apply f_pre to the leaf node but not f_post
  4) remove f_leaf and add extra parameter to `f_post/f_pre`, `leaf=TRUE`

I am somewhat indifferent between 1 and 2. Though 2 seems a bit cleaner conceptually and the default penalty of `is.list` is trivial.